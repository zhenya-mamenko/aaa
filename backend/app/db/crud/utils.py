import json
import re

from pydantic import BaseModel as _BaseModel

from app.db.sqlite import (
    execute_query as _execute_query,
    get_database_conn as _get_database_conn,
    read_query as _read_query,
)


def create_parameters_from_model(
    model: _BaseModel,
    /,
    exclude_fields: tuple[str] | None = None,
    only_fields: tuple[str] | None = None,
    get_max_data: tuple | None = None,
) -> tuple:
    """
    Create SQL parameters from a Pydantic model.

    Args:
        model (_BaseModel): The Pydantic model instance.
        exclude_fields (tuple[str] | None, optional): Fields to exclude from the model. Defaults to None.
        only_fields (tuple[str] | None, optional): Fields to include from the model. Defaults to None.
        get_max_data (tuple | None, optional): Tuple containing table name and field to get max value. Defaults to None.

    Returns:
        tuple: A tuple containing the SQL template, fields, parameters, and optionally a max query.
    """

    exclude_fields = exclude_fields or ()
    model_dump = {
        k: json.dumps(v) if type(v).__name__ in ["list", "dict"] else v for k, v in model.model_dump().items()
    }

    if only_fields:
        fields = [field for field in only_fields if field in model_dump]
        if len(fields) != len(only_fields):
            raise ValueError(f"Fields={set(fields) ^ set(only_fields)} are not in the model.")
    else:
        fields = [field for field, value in model_dump.items() if field not in exclude_fields and value is not None]
    parameters = tuple([model_dump[field] for field in fields])
    template = ", ".join(["?" for _ in parameters])

    if get_max_data:
        table_name, field = get_max_data
        if field in fields:
            raise ValueError(f"Field {field} must be excluded from the model.")
        fields.append(field)
        template += f", (SELECT COALESCE(MAX({field}), 0) + 1 FROM {table_name})"
        max_query = f"SELECT MAX({field}) as {field} FROM {table_name};"
        return template, fields, parameters, max_query

    return template, fields, parameters


def insert_data_from_model(
    *, table_name: str, id_field: tuple | str, model: _BaseModel, auto_increment_id: bool = True
) -> int | str | None:
    """
    Insert data into a table from a Pydantic model.

    Args:
        table_name (str): The name of the table to insert data into.
        id_field (tuple | str): The field(s) representing the ID.
        model (_BaseModel): The Pydantic model instance.
        auto_increment_id (bool, optional): Whether the ID is auto-incremented. Defaults to True.

    Returns:
        int | str | None: The ID of the inserted row if auto_increment_id is True, otherwise None.
    """

    if auto_increment_id:
        if isinstance(id_field, tuple):
            raise ValueError("auto_increment_id must be False if id_field is a tuple.")
        template, fields, parameters, q = create_parameters_from_model(model, get_max_data=(table_name, id_field))
    else:
        template, fields, parameters = create_parameters_from_model(model)

    with _get_database_conn() as conn:
        query = f"""
            INSERT INTO {table_name} (
                {", ".join(fields)}
            ) VALUES (
                {template}
            );
        """
        _execute_query(query, parameters=parameters, conn=conn)
        if auto_increment_id:
            id = _read_query(q, conn=conn)[0][id_field]
            return id

    return None


def update_data_from_model(*, table_name: str, id_field: tuple | str, model: _BaseModel) -> bool:
    """
    Update data in a table from a Pydantic model.

    Args:
        table_name (str): The name of the table to update data in.
        id_field (tuple | str): The field(s) representing the ID.
        model (_BaseModel): The Pydantic model instance.

    Returns:
        bool: True if the update was successful, False otherwise.
    """

    if isinstance(id_field, str):
        id_field = (id_field,)

    _, fields, parameters = create_parameters_from_model(model, exclude_fields=id_field)

    query = f"""
        UPDATE {table_name}
        SET
            {", ".join([f"{field} = ?" for field in fields])}
        WHERE
            {' AND '.join([f'{f} = ?' for f in id_field])};
    """
    parameters += tuple([getattr(model, f) for f in id_field])
    result = _execute_query(query, parameters=parameters)
    return result == 1


def delete_data_by_id(*, table_name: str, id_field: tuple | str, id_value: int | str | tuple) -> bool:
    """
    Delete data from a table by ID.

    Args:
        table_name (str): The name of the table to delete data from.
        id_field (tuple | str): The field(s) representing the ID.
        id_value (int | str | tuple): The value(s) of the ID to delete.

    Returns:
        bool: True if the deletion was successful, False otherwise.
    """

    if not isinstance(id_field, tuple):
        id_field=(id_field, )
    query = f"""
        DELETE FROM {table_name}
        WHERE
            {' AND '.join([f'{f} = ?' for f in id_field])};
    """
    if not isinstance(id_value, tuple):
        id_value=(id_value, )
    result = _execute_query(query, parameters=id_value)

    return result == 1


def extract_create_stmts_objects(sql_file_path: str, object_type: str) -> list[str]:
    """
    Extract the CREATE statements objects from a SQL file.

    Args:
        sql_file_path (str): The path to the SQL file.
        object_type (str): The type of objects to extract from the SQL file.

    Returns:
        list: A list of dictionaries containing the objects.
    """

    try:
        with open(sql_file_path, "r") as f:
            sql = f.read()
    except FileNotFoundError:
        raise FileNotFoundError(f"File not found: {sql_file_path}")

    re_template = fr"CREATE\s+{object_type}\s+(IF\s+NOT\s+EXISTS\s+)*(\S+)"
    pattern = re.compile(re_template)
    result = []
    for m in pattern.finditer(sql):
        result.append(m.group(2))

    return result
