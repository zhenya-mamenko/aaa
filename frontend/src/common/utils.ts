import { type AppElement } from "@/types/common";
import { elements } from "@/common/const";


export const isPlainObject: Function = (o: any) => Boolean(
  o && o.constructor && o.constructor.prototype && o.constructor.prototype.hasOwnProperty("isPrototypeOf")
);


export const flattenObject: Function = (obj: any, keys = []) => {
  return Object.keys(obj).reduce((acc: any, key: any) => {
    return Object.assign(acc, isPlainObject(obj[key])
      ? flattenObject(obj[key], keys.concat(key))
      : { [keys.concat(key).join(".")]: obj[key] }
    )
  }, {})
}


export const loadParams: Function = (key: string) => {
  const savedParams = localStorage.getItem(key)
  if (savedParams) {
    return JSON.parse(savedParams)
  }
  return null
}


export const saveParams: Function = (key: string, params: any) => {
  localStorage.setItem(key, JSON.stringify(params))
}


export const deleteParams: Function = (key: string) => {
  localStorage.removeItem(key)
}


export const removeReactivity: Function = (o: any) => {
  return JSON.parse(JSON.stringify(o))
}

export const findElement: Function = (key: string): AppElement | undefined => {
  return elements.find((element: AppElement) => element.key === key);
}


// String.formatUnicorn polyfill from https://stackoverflow.com/a/18234317
export const format: Function = function(str: string, ...rest: any[]): string {
    if (rest.length) {
      var t = typeof rest[0];
      var key;
      var args = ("string" === t || "number" === t) ? Array.prototype.slice.call(rest) : rest[0];
      for (key in args) {
        str = str.replace(new RegExp("\\{" + key + "\\}", "gi"), args[key]);
      }
    }
    return str;
  };


export const parseAxiosError: Function = (error: any): string | any[] => {
  if (error.response?.data?.detail !== undefined) {
    let message = "";
    const detail = error.response.data.detail;
    if (typeof detail === "string" && detail.length > 0) {
      message = `messages.backend.${detail.split(":")[0].replace(/ /g, "_")}`;
    }
    return [
      "messages.backend.request_error",
      error.response.status,
      message,
    ];
  } else if (!error.response) {
    return "messages.backend.no_reponse";
  } else {
    return error.message || "messages.backend.unknown_error";
  }
}


export const calculateReplenishment: Function = (
  list: any[],
  summa: number,
  percentileKey: string = "structure_percentile",
  resultKey: string = "calculated"
): void => {
  list.forEach(item => {
    (item as any)[resultKey] = Math.round(summa * item[percentileKey] / 1000);
  });
}


export const calculateRebalancing: Function = (
  list: any[],
  summa: number,
  amountKey: string = "amount",
  percentileKey: string = "structure_percentile",
  resultKey: string = "calculated"
): void => {
  const total = list.reduce((acc, item) => acc + item[amountKey], 0) + summa * 100;
  list.forEach(item => {
    (item as any)[resultKey] = Math.round(((total * item[percentileKey] / 1000) - item[amountKey]) / 100);
  });
}
