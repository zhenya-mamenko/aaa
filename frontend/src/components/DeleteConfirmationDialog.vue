<template>
  <v-dialog
    max-width="600px"
    v-model="opened"
  >
    <v-card
      class="pa-3"
    >
      <v-card-title
        class="ms-2"
        data-testid="delete-confirmation-dialog-title"
      >
        {{ t("confirmations.title") }}
      </v-card-title>
      <v-card-text>
        <div
          class="mb-2"
          data-testid="delete-confirmation-dialog-text"
        >
          {{ text }}
        </div>
        <div
          class="mb-3"
          data-testid="delete-confirmation-dialog-confirm-text"
        >
          {{ t("confirmations.delete_confirm") }}
        </div>
        <v-text-field
          data-testid="delete-confirmation-dialog-input"
          density="compact"
          hide-details
          required
          v-model="confirmationString"
        >
          <template v-slot:label>
            delete <span class="ms-1 warning">*</span>
          </template>
        </v-text-field>
      </v-card-text>
      <v-card-actions
        class="pe-4 me-2"
      >
        <v-spacer />
        <v-btn
          color="primary"
          data-testid="delete-confirmation-dialog-delete-button"
          :disabled="confirmationString !== 'delete'"
          @click="$emit('delete')"
        >
          {{ t("buttons.delete") }}
        </v-btn>
        <v-btn
          @click="$emit('close')"
        >
          {{ t("buttons.cancel") }}
        </v-btn>
      </v-card-actions>
    </v-card>
  </v-dialog>
</template>

<script lang="ts">
import { t } from "@/common/locale";
import { defineComponent } from "vue";


export default defineComponent({
  computed: {
    t() {
      return t;
    },
  },
  emits: ["close", "delete"],
  props: {
    text: {
      type: String,
      required: true,
    },
    opened: {
      type: Boolean,
      required: true,
    },
  },
  data() {
    return {
      confirmationString: "",
    };
  },
  watch: {
    opened(value) {
      if (value) {
        this.confirmationString = "";
      }
    },
  },
});
</script>
