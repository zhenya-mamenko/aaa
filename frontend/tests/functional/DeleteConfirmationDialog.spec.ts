import { mountComponent, cardStubs, simpleStubs } from "./utils";
// @ts-ignore
import DeleteConfirmationDialog from "@/components/DeleteConfirmationDialog.vue";

jest.mock("@/common/locale", () => ({
  t: (key: string) => key,
}));

describe("DeleteConfirmationDialog", () => {

  function mountDialog(props: any, data: any) {
    return mountComponent(
      DeleteConfirmationDialog,
      {
        simpleStubs,
        stubs: [...cardStubs, "v-dialog"],
      },
      {
        opened: true,
        text: "Test",
        ...props,
      },
      {
        confirmationString: "test",
        ...data,
      }
    );
  }

  it("Should render", () => {
    const wrapper = mountDialog({}, {});

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("Should cointain the right texts", () => {
    const wrapper = mountDialog({}, {});

    const title = wrapper.get("[data-testid='delete-confirmation-dialog-title']");
    expect(title.text()).toEqual("confirmations.title");

    const text = wrapper.get("[data-testid='delete-confirmation-dialog-text']");
    expect(text.text()).toEqual("Test");

    const confirmText = wrapper.get("[data-testid='delete-confirmation-dialog-confirm-text']");
    expect(confirmText.text()).toEqual("confirmations.delete_confirm");
  });

  it("Should enable/disable the delete button", () => {
    const wrapper = mountDialog({}, {});

    const input = wrapper.get("[data-testid='delete-confirmation-dialog-input']");
    expect(input.attributes("modelvalue")).toEqual("test");

    const disabled = wrapper.get("[data-testid='delete-confirmation-dialog-delete-button']");
    expect(disabled.attributes("disabled")).toEqual("true");

    const wrapper2 = mountDialog({},
      {
        confirmationString: "delete",
      }
    );

    const input2 = wrapper2.get("[data-testid='delete-confirmation-dialog-input']");
    expect(input2.attributes("modelvalue")).toEqual("delete");

    const enabled = wrapper2.get("[data-testid='delete-confirmation-dialog-delete-button']");
    expect(enabled.attributes("disabled")).toEqual("false");
  });

});
