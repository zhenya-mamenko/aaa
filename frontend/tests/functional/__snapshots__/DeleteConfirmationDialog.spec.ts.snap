// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeleteConfirmationDialog Should render 1`] = `
"<div max-width="600px" modelvalue="true">
  <div class="pa-3">
    <div class="ms-2" data-testid="delete-confirmation-dialog-title">confirmations.title</div>
    <div>
      <div class="mb-2" data-testid="delete-confirmation-dialog-text">Test</div>
      <div class="mb-3" data-testid="delete-confirmation-dialog-confirm-text">confirmations.delete_confirm</div>
      <v-text-field-stub data-testid="delete-confirmation-dialog-input" density="compact" hide-details="" required="" modelvalue="test"></v-text-field-stub>
    </div>
    <div class="pe-4 me-2">
      <v-spacer-stub></v-spacer-stub>
      <v-btn-stub color="primary" data-testid="delete-confirmation-dialog-delete-button" disabled="true"></v-btn-stub>
      <v-btn-stub></v-btn-stub>
    </div>
  </div>
</div>"
`;
